### PR TITLE
Updating submodules and pace refactor

### PR DESCRIPTION
**Pull request by fmalatino**
**Description**
This PR updates the `NDSL`, `pyFV3`, and `pySHiELD` submodules to most recent commits, reflecting the standardized import methods.

**How Has This Been Tested?**
Tested using the currently implemented unit tests in `tests/main`

**Checklist:**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules
- [x] New check tests, if applicable, are included

### CODE REVIEW

File: driver/pace/driver/__init__.py [hunk 1]
```diff
@@ -8,11 +8,30 @@
     ReaderCommConfig,
     WriterCommConfig,
 )
-from .diagnostics import Diagnostics, DiagnosticsConfig
+from .diagnostics import (
+    Diagnostics,
+    DiagnosticsConfig,
+    MonitorDiagnostics,
+    NullDiagnostics,
+    ZSelect,
```
Comment (oelbert) [inline driver/pace/driver/__init__.py]: What's the logic behind adding these here?

File: driver/pace/driver/run.py [hunk 2]
```diff
@@ -6,8 +6,7 @@
 import yaml
 
 from ndsl.logging import AVAILABLE_LOG_LEVELS, ndsl_log
-
-from .driver import Driver, DriverConfig
+from pace.driver.driver import Driver, DriverConfig
```
Comment (oelbert) [inline driver/pace/driver/run.py]: What's the reason for this change?

**Review by oelbert — COMMENTED** [review]
Two quick questions on the driver init and relative imports, just because it looks a little awkward but nothing blocking

File: driver/pace/driver/__init__.py (hunk 1 above)
Comment (fmalatino) [reply driver/pace/driver/__init__.py]: I was just copying what was done in NDSL, with exposing without rationale, and then un-exposing when it is deemed unnecessary.

File: driver/pace/driver/run.py (hunk 2 above)
Comment (fmalatino) [reply driver/pace/driver/run.py]: I am following the style that was set forth in how I handled imports in the submodules. When I am within a package and I am referencing something internal to the package, the absolute path is used, when outside the package the relative path is used.

Comment (oelbert) [reply driver/pace/driver/run.py]: Yeah that makes sense, it just looks funny here

**Review by oelbert — APPROVED** [review]
👍

**Review by bensonr — COMMENTED** [review]
As I look at the driver/pace/driver code I have a question about what we are building with Pace.  If Pace is designed to be the model and we will be adding other components (land, ice, ocean, waves, etc.) within, I'm not sure it makes senes to treat the driver as a submodule.  However, if we envision Pace being say the atmospheric component within a larger module, it does make sense to leave the submodule treatment in place.

**Comment by fmalatino** [comment]
> As I look at the driver/pace/driver code I have a question about what we are building with Pace. If Pace is designed to be the model and we will be adding other components (land, ice, ocean, waves, etc.) within, I'm not sure it makes senes to treat the driver as a submodule. However, if we envision Pace being say the atmospheric component within a larger module, it does make sense to leave the submodule treatment in place.

@oelbert ?

**Comment by oelbert** [comment]
My understanding has been that Pace is solely an atmospheric model and any ocean or land model would be separate

**Comment by lharris4** [comment]
Currently Pace is intended to be an atmosphere model, possibly containing a
land component. Ocean and ice are separate from Pace's scope.

Lucas

File: Makefile [hunk 3]
```diff
@@ -142,8 +142,8 @@ test_mpi_54rank:
 	mpirun -n 54 $(MPIRUN_ARGS) python3 -m mpi4py -m pytest tests/mpi_54rank
```
Comment (bensonr) [inline Makefile:142]: Will we always be using mpirun here or is there a chance it could be another invoker?

File: driver/pace/README.md [hunk 4]
```diff
@@ -1,17 +1,17 @@
-# pace-driver
+# pace
 
 This package provides command-line routines to run the Pace model, and utilities to write model driver scripts.
 
-We suggest reading the code in the examples directory, or taking a look at `pace/driver/run.py` to see how the main entrypoint for this package works.
+We suggest reading the code in the examples directory, or taking a look at `driver/pace/driver/run.py` to see how the main entrypoint for this package works.
```
Comment (bensonr) [inline driver/pace/README.md]: Did the structure change to driver/pace/run.py or is it still driver/pace/driver/run.py?

**Review by bensonr — COMMENTED** [review]
We are using srun in the .jenkins code.  Should we use a macro that can be either srun or mpirun as we shouldn't always assume a slurm environment?

Is all of the documentation that is NDSL, pyFV3, or pySHiELD specific also duplicated in those repos?  Or better yet, it should live in those repos and become available via the submodule includes.

The driver/pace structure still seems redundant to me.  Perhaps it is my lack of Python semantics, but we are in the Pace repo and have a driver/.  Why do I need to yet again remind you this is pace by having driver/pace/?   Is it better to have pace/ or pace/driver so that all references are pace.X or pace.driver.X?

**Comment by oelbert** [comment]
> The driver/pace structure still seems redundant to me. Perhaps it is my lack of Python semantics, but we are in the Pace repo and have a driver/. Why do I need to yet again remind you this is pace by having driver/pace/? Is it better to have pace/ or pace/driver so that all references are pace.X or pace.driver.X?

pace/driver/pace/driver/* structure is definitely a relic of Pace's development history, and we want to resolve that. Do you think we should do it here or in a separate PR?

**Comment by bensonr** [comment]
In response to [this comment](https://github.com/NOAA-GFDL/pace/pull/75#issuecomment-2059716805), this PR is already renaming things from driver/pace/driver.  Instead of doing it yet again, I was hoping we could get the structure set in one step.  As always, I want the approach with the least chance of breaking things.

**Comment by fmalatino** [comment]
> In response to [this comment](https://github.com/NOAA-GFDL/pace/pull/75#issuecomment-2059716805), this PR is already renaming things from driver/pace/driver. Instead of doing it yet again, I was hoping we could get the structure set in one step. As always, I want the approach with the least chance of breaking things.

Currently the method of setting up the `pace` package differs from its submodules in that it uses a requirements file for specifying the requirements for the `pip install` of the package, as opposed to passing them through a `setup.py` or other build backend method. Within the requirements file we specify a need for an editable install of the submodules and pace, which will search for a `setup.py` file in each of the directories specified. Each of the `setup.py` files look for a namespace package matching the specified name(s) in the call to the `find_namespace_packages`. These namespace packages must be contained within a directory alongside the `setup.py` file. 
Proposed in PR:
```
driver/
├── pace/      
│   ├── __init__.py      
│   └── driver.py
│   └── other.py            
└── setup.py
```
Current:
```
driver/
├── pace/ 
│     ├── driver/      
│          ├── __init__.py      
│          └── driver.py
│          └── other.py            
└── setup.py
```
I chose `driver` to serve as the outer directory to contain the `pace` package and the `setup.py` with the intention that once installed the user would then use the package like:
```
import pace

driver = pace.Driver(...)
```

As opposed to what existed previously:

```
import pace.driver

driver = pace.driver.Driver(...)
```
We can change the use of `setup.py` to include all of the desired requirements and make calls to install the submodules as well, which will then treat the cloned repo as the outer directory containing the pace package alongside its `setup.py`:
```
cloned_pace/
├── pace/      
│   ├── __init__.py      
│   └── driver.py
│   └── other.py            
└── setup.py
└── other_mods/
```
This will however not use the requirements file method, and will require some minor changes elsewhere, like in the github workflows, etc. I have tested both of these methods and went with the former as it keeps prevents a need to change the other files. If the latter is preferred the changes can be easily made. Also, I am not attached to the name `driver` for the outer directory, only using it as it made it easy to implement.

**Comment by fmalatino** [comment]
@bensonr and @oelbert 

I have made changes to the `requirements_dev.txt` file which should allow for pace to be installed without an outer directory (aside from it being contained in the cloned repo).

File: pace/README.md [hunk 5]
```diff
@@ -1,17 +1,17 @@
-# pace-driver
+# pace
 
 This package provides command-line routines to run the Pace model, and utilities to write model driver scripts.
 
-We suggest reading the code in the examples directory, or taking a look at `pace/driver/run.py` to see how the main entrypoint for this package works.
+We suggest reading the code in the examples directory, or taking a look at `driver/pace/driver/run.py` to see how the main entrypoint for this package works.
```
Comment (bensonr) [inline pace/README.md]: I think this should be `pace/run.py`.

File: setup.py [hunk 6]
```diff
@@ -33,11 +36,9 @@
         "Programming Language :: Python :: 3.9",
     ],
     install_requires=requirements,
-    setup_requires=setup_requirements,
-    tests_require=test_requirements,
-    name="pace-driver",
+    name="pace",
```
Comment (bensonr) [inline setup.py:39]: Do we need to update more of this info here in the setup block here?

**Review by bensonr — APPROVED** [review]
Thanks for taking my suggestions regarding structure.  You've done a great job making it work.  There are two minor nits that could be saved for a future issue/PR.

File: setup.py (hunk 6 above)
Comment (fmalatino) [reply setup.py:39]: I think this may be a good issue as I am not sure what would be appropriate to retain relating to the AI2 work.

File: pace/README.md (hunk 5 above)
Comment (fmalatino) [reply pace/README.md]: Amended in 15fc2955

File: Makefile (hunk 3 above)
Comment (oelbert) [reply Makefile:142]: A change here would be prompted by a change in the docker image

**Review by oelbert — APPROVED** [review]
Phew, that's a lot of changes!

File: .github/workflows/main_unit_tests.yml [hunk 7]
```diff
@@ -0,0 +1,31 @@
+name: "Main unit tests"
```
Comment (mlee03) [inline .github/workflows/main_unit_tests.yml:1]: Is this workflow file needed?

Comment (fmalatino) [reply .github/workflows/main_unit_tests.yml:1]: Are the tests contained within that directory also tested in CI? If so, it can be removed.
